src/QMCWaveFunctions/Fermion/DiracDeterminant.h

```diff
@@ -97,6 +97,9 @@ class DiracDeterminant : public DiracDeterminantBase
 
   ValueType ratioGrad(ParticleSet& P, int iat, GradType& grad_iat);
   GradType evalGrad(ParticleSet& P, int iat);
+
+  void prepareIonDerivs();
```

Comment:
This function is expected to be private and only called within DiracDeterminant class.
Calling from outside and passing through all wavefunction hierarchy is bad.
Also it is better to call it resizeScratchObjectsForIonDerivs.

Reply:
This function is where I got hamstrung by the current API.  Right now, the evaluateGradSource functions take an ion index and return a 3d gradient associated with that ion's ionic derivatives.  So I could pull specific force gradients out, or I could compute the gradients for all N ions.  If I want to do the latter, I want to allocate and populate temporary arrays once at the beginning of all these computations if the electron particle set has changed.  Having this function here was in anticipation of that, although I could probably just add a function "evaluateGradSource" function that computes ALL the force derivatives, and does the allocation within itself.  I think I might do that...

Reply:
memory allocation cost is paid only once even if you call resize infinite times. That is why I insist moving it into the evaluation function. Is it a concern for you?

My understanding is that evaluateGradSource should compute ALL force derivatives. It doesn't make sense to me just do one and it stops us writing very efficient codes.
Is evalGradSourcep used? If not, could you delete it?

Reply:
The resize itself is not a problem.  It's if I have to do stuff other than resize.  But I think adding another function is the right way to handle this case.  Your suggestion to add a resize to the current function is the best solution.